apps/src/lib/client/tendermint_rpc_types.rs

```diff
@@ -34,65 +35,69 @@ pub struct TxResponse {
-            .unwrap_or_else(tx_error!())
-            .iter()
-            .enumerate()
-            .find(|(_, hash)| hash == &tx_hash)
-            .unwrap_or_else(tx_error!());
-        let info = events[&format!("{evt_key}.info")][index].clone();
-        let log = events[&format!("{evt_key}.log")][index].clone();
-        let height = events[&format!("{evt_key}.height")][index].clone();
-        let code = events[&format!("{evt_key}.code")][index].clone();
-        let gas_used = events[&format!("{evt_key}.gas_used")][index].clone();
-        let initialized_accounts = events
-            [&format!("{evt_key}.initialized_accounts")]
-            .get(index)
-            .as_ref()
-            .map(|initialized_accounts| {
-                serde_json::from_str(initialized_accounts).unwrap()
-            })
-            .unwrap_or_else(|| {
-                eprintln!(
-                    "Tendermint omitted one of the expected indices in events"
-                );
-                Vec::new()
-            });
-        TxResponse {
+
+        let hash = event
+            .get("hash")
+            .ok_or_else(|| missing_field_err("hash"))?
+            .clone();
+        let info = event
+            .get("info")
+            .ok_or_else(|| missing_field_err("info"))?
+            .clone();
+        let log = event
+            .get("log")
+            .ok_or_else(|| missing_field_err("log"))?
+            .clone();
+        let height = event
+            .get("height")
+            .ok_or_else(|| missing_field_err("height"))?
+            .clone();
+        let code = event
+            .get("code")
+            .ok_or_else(|| missing_field_err("code"))?
+            .clone();
+        let gas_used = event
+            .get("gas_used")
+            .ok_or_else(|| missing_field_err("gas_used"))?
+            .clone();
+        let initialized_accounts = event
+            .get("initialized_accounts")
+            .map(String::as_str)
+            // TODO: fix finalize block, to return initialized accounts,
+            // even when we reject a tx?
+            .or(Some("[]"))
+            // NOTE: at this point we only have `Some(vec)`, not `None`
+            .ok_or_else(|| unreachable!())
+            .and_then(|initialized_accounts| {
+                serde_json::from_str(initialized_accounts)
+                    .map_err(|err| format!("JSON decode error: {err}"))
+            })?;
+
+        Ok(TxResponse {
+            hash,
             info,
             log,
             height,
             code,
             gas_used,
             initialized_accounts,
-            hash: tx_hash.to_string(),
-        }
+        })
+    }
+}
+
+impl TxResponse {
+    /// Convert an [`Event`] to a [`TxResponse`], or error out.
+    pub fn from_event(event: Event) -> Self {
+        event.try_into().unwrap_or_else(|err| {
+            eprintln!("Error fetching TxResponse: {err}");
+            safe_exit(1);
+        })
     }
 }
```

Comment:
To do later - https://github.com/anoma/anoma/issues/1049 - these `safe_exit`s need to be removed

Reply:
This sounds like an issue with the test runner. It is supposed to watch the exit status and return an error if it gets a non-zero value.

Reply:
It would be good if the test runner could catch that, but it's still bad that we have these functions maybe taking a decision to terminate the process